Don't hold idMDMU lock while calling UpdateIdentities from InjectLabels

codablock opened the pull request:
UpdateIdentities needs to lock SelectorCache.mutex, which might be held
by SelectorCache.AddFQDNSelector while endpoints/policies are being
regenerated. If in that situation idMDMU is locked already, AddFQDNSelector
will deadlock while calling ipcache.GetIDMetadataByIP deep inside the
call chain.

Example stacktraces of 2 goroutines that show the deadlock:

```
goroutine 236 [semacquire, 61 minutes]:
sync.runtime_SemacquireMutex(0x5000107, 0x0, 0xffffffffffffffff)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc0004d13b0)
	/usr/local/go/src/sync/mutex.go:138 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
sync.(*RWMutex).Lock(0xc001365f58)
	/usr/local/go/src/sync/rwmutex.go:111 +0x36
github.com/cilium/cilium/pkg/policy.(*SelectorCache).UpdateIdentities(0xc0004d13b0, 0xc001f0dbf0, 0x0, 0x8)
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:949 +0x65
github.com/cilium/cilium/pkg/ipcache.InjectLabels({0x27fe323, 0xf}, {0x2bfdbe0, 0xc0004d13b0}, {0x7f1837d6d080, 0xc0009e0e20})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:179 +0xa53
github.com/cilium/cilium/pkg/ipcache.(*IPCache).TriggerLabelInjection.func1({0xc0013674d0, 0xc001367190})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:550 +0x31
github.com/cilium/cilium/pkg/controller.(*Controller).runController(0xc0005a4360)
	/go/src/github.com/cilium/cilium/pkg/controller/controller.go:206 +0x1cb
created by github.com/cilium/cilium/pkg/controller.(*Manager).updateController
	/go/src/github.com/cilium/cilium/pkg/controller/manager.go:111 +0xb67

goroutine 2578413 [semacquire, 61 minutes]:
sync.runtime_SemacquireMutex(0xe, 0x0, 0xc000c76000)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:63
github.com/cilium/cilium/pkg/ipcache.GetIDMetadataByIP({0xc001c5b830, 0x17})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:68 +0x68
github.com/cilium/cilium/pkg/ipcache.AllocateCIDRs({0xc0018ca660, 0x4, 0xc00041fce0}, 0x0)
	/go/src/github.com/cilium/cilium/pkg/ipcache/cidr.go:62 +0x20a
github.com/cilium/cilium/pkg/ipcache.AllocateCIDRsForIPs({0xc00009bc20, 0x5, 0xc003a46e80}, 0x37)
	/go/src/github.com/cilium/cilium/pkg/ipcache/cidr.go:99 +0x2f
github.com/cilium/cilium/daemon/cmd.cachingIdentityAllocator.AllocateCIDRsForIPs({0xc00041fc70}, {0xc00009bc20, 0xc00278f510, 0x3}, 0x40e3cb)
	/go/src/github.com/cilium/cilium/daemon/cmd/identity.go:113 +0x2a
github.com/cilium/cilium/pkg/policy.(*fqdnSelector).allocateIdentityMappings(0xc002698370, {0x7f1837c69138, 0xc000ac05a0}, 0x25)
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:491 +0x258
github.com/cilium/cilium/pkg/policy.(*SelectorCache).AddFQDNSelector(0xc0004d13b0, {0x2bfdbc0, 0xc002851a00}, {{0x0, 0x0}, {0xc004c7ec60, 0xa}})
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:820 +0x40b
github.com/cilium/cilium/pkg/policy.(*L4Filter).cacheFQDNSelector(0xc002851a00, {{0x0, 0xc00000c180}, {0xc004c7ec60, 0xc000e6eb10}}, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:479 +0x65
github.com/cilium/cilium/pkg/policy.(*L4Filter).cacheFQDNSelectors(0xc0018ca280, {0xc000988380, 0xe, 0x2c732c0}, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:474 +0x8f
github.com/cilium/cilium/pkg/policy.createL4Filter({0x2c732c0, 0xc0018ca060}, {0xc0018ca280, 0x1, 0x1}, {0x2c24098, 0xc0029a9740}, {{0x2bd17b2, 0x1}, {0x27d4b6a, ...}}, ...)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:574 +0x398
github.com/cilium/cilium/pkg/policy.createL4EgressFilter(...)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:697
github.com/cilium/cilium/pkg/policy.mergeEgressPortProto({0x2c732c0, 0xc0018ca060}, 0x0, {0xc0018ca280, 0xc0025cfe38, 0x44cd45}, {0x2c24098, 0xc0029a9740}, {{0x2bd17b2, 0x1}, ...}, ...)
	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:773 +0x108
github.com/cilium/cilium/pkg/policy.mergeEgress({0x2c732c0, 0xc0018ca060}, 0xc001f0e3f0, {0xc0018ca280, 0x1, 0x1}, {0x2c2a268, 0x4875340}, {0x2c2a218, 0x4875340}, ...)
	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:678 +0x35c
github.com/cilium/cilium/pkg/policy.(*rule).resolveEgressPolicy(0xc0023fb860, {0x2c732c0, 0xc0018ca060}, 0xc001f0e3f0, 0xc002790308, 0xc0029a9620, {0x0, 0x0, 0x0}, {0x0, ...})
	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:809 +0x84f
github.com/cilium/cilium/pkg/policy.ruleSlice.resolveL4EgressPolicy({0xc00299af40, 0x6, 0xb7}, {0x2c732c0, 0xc0018ca060}, 0xc001f0e3f0)
	/go/src/github.com/cilium/cilium/pkg/policy/rules.go:102 +0x470
github.com/cilium/cilium/pkg/policy.(*Repository).resolvePolicyLocked(0xc0004d1490, 0xc001fcdbc0)
	/go/src/github.com/cilium/cilium/pkg/policy/repository.go:704 +0x531
github.com/cilium/cilium/pkg/policy.(*PolicyCache).updateSelectorPolicy(0xc000852a98, 0xc001fcdbc0)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:119 +0x14c
github.com/cilium/cilium/pkg/policy.(*PolicyCache).UpdatePolicy(...)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:153
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regeneratePolicy(0xc000ee3880)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:230 +0x22b
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0xc000ee3880, 0xc000bff400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:815 +0x2dd
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc000ee3880, 0xc000bff400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:584 +0x19d
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc000ee3880, 0xc000bff400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:405 +0x7b3
github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0xc003a950d0, 0xc001fcd620)
	/go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x32c
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x13b
sync.(*Once).doSlow(0xc000c25901, 0x43f325)
	/usr/local/go/src/sync/once.go:68 +0xd2
sync.(*Once).Do(...)
	/usr/local/go/src/sync/once.go:59
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0xc00304ae40)
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x45
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x7b
```

The deadlock was discovered in a self-compiled version (branch v1.11) of cilium. I assume the fix needs to be at least backported into v.1.11, maybe also into v1.10.


```release-note
Fix deadlock with kube-apiserver policy matching feature
```

codablock committed:
Don't hold idMDMU lock while calling UpdateIdentities from InjectLabels

UpdateIdentities needs to lock SelectorCache.mutex, which might be held
by SelectorCache.AddFQDNSelector while endpoints/policies are being
regenerated. If in that situation idMDMU is locked already, AddFQDNSelector
will deadlock while calling ipcache.GetIDMetadataByIP deep inside the
call chain.

Example stacktraces of 2 goroutines that show the deadlock:

```
goroutine 236 [semacquire, 61 minutes]:
sync.runtime_SemacquireMutex(0x5000107, 0x0, 0xffffffffffffffff)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc0004d13b0)
	/usr/local/go/src/sync/mutex.go:138 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
sync.(*RWMutex).Lock(0xc001365f58)
	/usr/local/go/src/sync/rwmutex.go:111 +0x36
github.com/cilium/cilium/pkg/policy.(*SelectorCache).UpdateIdentities(0xc0004d13b0, 0xc001f0dbf0, 0x0, 0x8)
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:949 +0x65
github.com/cilium/cilium/pkg/ipcache.InjectLabels({0x27fe323, 0xf}, {0x2bfdbe0, 0xc0004d13b0}, {0x7f1837d6d080, 0xc0009e0e20})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:179 +0xa53
github.com/cilium/cilium/pkg/ipcache.(*IPCache).TriggerLabelInjection.func1({0xc0013674d0, 0xc001367190})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:550 +0x31
github.com/cilium/cilium/pkg/controller.(*Controller).runController(0xc0005a4360)
	/go/src/github.com/cilium/cilium/pkg/controller/controller.go:206 +0x1cb
created by github.com/cilium/cilium/pkg/controller.(*Manager).updateController
	/go/src/github.com/cilium/cilium/pkg/controller/manager.go:111 +0xb67

goroutine 2578413 [semacquire, 61 minutes]:
sync.runtime_SemacquireMutex(0xe, 0x0, 0xc000c76000)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:63
github.com/cilium/cilium/pkg/ipcache.GetIDMetadataByIP({0xc001c5b830, 0x17})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:68 +0x68
github.com/cilium/cilium/pkg/ipcache.AllocateCIDRs({0xc0018ca660, 0x4, 0xc00041fce0}, 0x0)
	/go/src/github.com/cilium/cilium/pkg/ipcache/cidr.go:62 +0x20a
github.com/cilium/cilium/pkg/ipcache.AllocateCIDRsForIPs({0xc00009bc20, 0x5, 0xc003a46e80}, 0x37)
	/go/src/github.com/cilium/cilium/pkg/ipcache/cidr.go:99 +0x2f
github.com/cilium/cilium/daemon/cmd.cachingIdentityAllocator.AllocateCIDRsForIPs({0xc00041fc70}, {0xc00009bc20, 0xc00278f510, 0x3}, 0x40e3cb)
	/go/src/github.com/cilium/cilium/daemon/cmd/identity.go:113 +0x2a
github.com/cilium/cilium/pkg/policy.(*fqdnSelector).allocateIdentityMappings(0xc002698370, {0x7f1837c69138, 0xc000ac05a0}, 0x25)
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:491 +0x258
github.com/cilium/cilium/pkg/policy.(*SelectorCache).AddFQDNSelector(0xc0004d13b0, {0x2bfdbc0, 0xc002851a00}, {{0x0, 0x0}, {0xc004c7ec60, 0xa}})
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:820 +0x40b
github.com/cilium/cilium/pkg/policy.(*L4Filter).cacheFQDNSelector(0xc002851a00, {{0x0, 0xc00000c180}, {0xc004c7ec60, 0xc000e6eb10}}, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:479 +0x65
github.com/cilium/cilium/pkg/policy.(*L4Filter).cacheFQDNSelectors(0xc0018ca280, {0xc000988380, 0xe, 0x2c732c0}, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:474 +0x8f
github.com/cilium/cilium/pkg/policy.createL4Filter({0x2c732c0, 0xc0018ca060}, {0xc0018ca280, 0x1, 0x1}, {0x2c24098, 0xc0029a9740}, {{0x2bd17b2, 0x1}, {0x27d4b6a, ...}}, ...)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:574 +0x398
github.com/cilium/cilium/pkg/policy.createL4EgressFilter(...)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:697
github.com/cilium/cilium/pkg/policy.mergeEgressPortProto({0x2c732c0, 0xc0018ca060}, 0x0, {0xc0018ca280, 0xc0025cfe38, 0x44cd45}, {0x2c24098, 0xc0029a9740}, {{0x2bd17b2, 0x1}, ...}, ...)
	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:773 +0x108
github.com/cilium/cilium/pkg/policy.mergeEgress({0x2c732c0, 0xc0018ca060}, 0xc001f0e3f0, {0xc0018ca280, 0x1, 0x1}, {0x2c2a268, 0x4875340}, {0x2c2a218, 0x4875340}, ...)
	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:678 +0x35c
github.com/cilium/cilium/pkg/policy.(*rule).resolveEgressPolicy(0xc0023fb860, {0x2c732c0, 0xc0018ca060}, 0xc001f0e3f0, 0xc002790308, 0xc0029a9620, {0x0, 0x0, 0x0}, {0x0, ...})
	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:809 +0x84f
github.com/cilium/cilium/pkg/policy.ruleSlice.resolveL4EgressPolicy({0xc00299af40, 0x6, 0xb7}, {0x2c732c0, 0xc0018ca060}, 0xc001f0e3f0)
	/go/src/github.com/cilium/cilium/pkg/policy/rules.go:102 +0x470
github.com/cilium/cilium/pkg/policy.(*Repository).resolvePolicyLocked(0xc0004d1490, 0xc001fcdbc0)
	/go/src/github.com/cilium/cilium/pkg/policy/repository.go:704 +0x531
github.com/cilium/cilium/pkg/policy.(*PolicyCache).updateSelectorPolicy(0xc000852a98, 0xc001fcdbc0)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:119 +0x14c
github.com/cilium/cilium/pkg/policy.(*PolicyCache).UpdatePolicy(...)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:153
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regeneratePolicy(0xc000ee3880)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:230 +0x22b
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0xc000ee3880, 0xc000bff400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:815 +0x2dd
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc000ee3880, 0xc000bff400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:584 +0x19d
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc000ee3880, 0xc000bff400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:405 +0x7b3
github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0xc003a950d0, 0xc001fcd620)
	/go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x32c
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x13b
sync.(*Once).doSlow(0xc000c25901, 0x43f325)
	/usr/local/go/src/sync/once.go:68 +0xd2
sync.(*Once).Do(...)
	/usr/local/go/src/sync/once.go:59
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0xc00304ae40)
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x45
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x7b
```


Signed-off-by: Alexander Block <ablock84@gmail.com>

christarazi reviewed (approved):
Thanks for the PR! That looks about right to me. I think we should wait until https://github.com/cilium/cilium/pull/18150 is merged as we might run into merge conflicts. 

I think it's alright for the `idMDMU` lock to be released before the ipcache lock is taken. I can't think of any TOCTOU implications. @joestringer I think we've touched on this while iterating on this feature.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -100,11 +100,11 @@ func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTr
 	)
 
 	idMDMU.Lock()
-	defer idMDMU.Unlock()
```
Looking at `InjectLabels()` in isolation, I think this would be OK because `InjectLabels()` is not intended to be reentrant, and there's only one caller anyway so the code inside doesn't need to protect against multiple callers calling in concurrently.

However, `InjectLabels()` has a corresponding opposite operation, `RemoveLabelsExcluded()` which has similar logic and these two will currently mutually exclude each other from concurrently manipulating the IdentityMetadata map + apply the changes resulting from those changes. This exclusion occurs using `idMDMU`, which would change with the current proposal in this PR. This adds additional complications to resolving the overall issue for two reasons:
1. Presumably if this issue is caused by policy calculation (potentially for any FQDN policy) happening at the same time as `InjectLabels()` (from node / k8s apiserver updates), similarly an equivalent issue could also occur if policy calculation happens at the same time that `RemoveLabelsExcluded()` occurs (this time, only from k8s apiserver updates). So while this solves one of the problems, there is still another similar issue in a related part of the code.
2. Regardless of the problem that is highlighted in the PR description,  it will now also make it possible for one goroutine to be handling e.g. a node update inside `InjectLabels()` while another goroutine handles a kube-apiserver endpoint update via `RemoveLabelsExcluded()` -> `removeLabelsFromIPs()`. This is not currently possible because the `idMDMU` mutex puts a critical section around the core logic for both of these functions. With the current proposal, if `InjectLabels()` runs slightly earlier, then it will grab the `idMDMU` lock and manipulate the IdentityMetadata map and calculate the changes that must subsequently occur, then release the lock, and then depending on the scheduling of goroutines on the system, could plausibly be delayed arbitrarily while the `RemoveLabelsExcluded()` call grabs the idMDMU, manipulates & calculates new state, and then issues the updates into the SelectorCache & IPCache. Depending on the scheduling, any of the possible state transitions could occur:
   a. All of `InjectLabels()` occurs, then all of `RemoveLabelsExcluded()`
   b. All of `RemoveLabelsExcluded()` occurs, then all of `InjectLabels()`
   c. The first part of `InjectLabels()` occurs here, then due to rescheduling of goroutines, all of `RemoveLabelsExcluded()` occurs, then the remainder of `InjectLabels()` occurs.

(2) starts to get a bit abstract, probably if the IPs are entirely different then there's no problem. But if the IPs collide between the two implementations, then rather than consistently first applying the changes to the `IdentityMetadata` map and then deriving the resulting state to apply to the `SelectorCache` / `IPCache` in order, it could end up mixing the order of operations and I'm not entirely convinced whether the code would end up consistently getting the SelectorCache+IPcache into the same state regardless of which conditions (a), (b), or (c) occur.

Broadly I see a couple of options that could resolve both of these issues: One is to ensure that both `InjectLabels()` and `RemoveLabelsExcluded()` occur inside the same goroutine, or the second would be to extend _some_ lock to cover the critical sections of both of these functions. My initial thought is "Well what about the `IPIdentityCache.mutex`?". Unfortunately, looking over the callstacks you provided, this will only defer the deadlock until slightly later since the other callstack will also attempt to grab the `IPIdentityCache.mutex` and push those changes down to the datapath while also holding the SelectorCache mutex. So `IPIdentityCache.mutex` is not an option.

joestringer replied:
For the immediate term, I wonder whether we can simply add a brand new mutex specifically for coordinating just `InjectLabels()` vs. `RemoveLabelsExcluded()`. Either function would first grab this `ApplyIDMDChanges` mutex before the `idMDMU`, and release this mutex only once all of the IdentityMetadata, SelectorCache and IPCache operations are completed. Then it should be safe to shorten the scope of the current IDMDU lock exactly the way that this patch is already proposing, and additionally we could apply a similar fix for the equivalent `RemoveLabelsExcluded()` -> `removeLabelsFromIPs()` code.

Longer term, we probably need to restructure the handling of events in and around the IPcache subsystem, and the way it interacts with the policy / `SelectorCache` layers to formalize the ordering of operations and provide better guarantees. But I don't have a fully formed idea around what that would look like just yet, and it's probably a bit unreasonable to immediately push for such a solution from you given that you've only recently started digging around in the code :-)

christarazi replied:
Good point. Thinking about this again, I didn't realize the `RemoveLabelsExcluded()` angle. 

From point (2) above, anything other than option (a) is undesirable (or vice versa where `RemoveLabelsExcluded()` completes first, then `InjectLabels()`; of course assuming that the K8s events are in order, but I think that's a safe assumption).

I agree in the immediate term a separate mutex makes sense which stitches together the two other mutexes, eliminating the gap in critical section, so-to-speak.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -148,6 +148,8 @@ func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTr
 			}
 		}
 	}
+	// Don't hold lock while calling UpdateIdentities, as it will otherwise run into a deadlock
```
(unrelated to this line)

I'm curious how you were able to reproduce this and whether there's an opportunity to introduce any tests to cover this? Clearly we didn't catch this the first time, and given that this is a lock coordination issue across multiple goroutines in separate subsystems, it can be a little tricky to identify purely through code review of specific areas.

codablock replied:
I encountered the deadlock in the same situation as described in #18352 (high load, many nodes starting up and getting scheduled with 100% IP use immediately due to many pending PODs). The symptom was that something (I don't remember what exactly) was suddenly stuck on many nodes and that listing endpoints via cilium cli was also hangling. I then used some go debugging tool (also forget the name of it...can't check atm) to dump all stack traces.  Based on the call stacks of I discovered the 2 offending locks, after which I started analysing the corresponding code.

I don't have a reliable reproducer unfortunately.

joestringer reviewed (changes requested):
Nice find!

I think there's other issues both adjacent to this issue and unfortunately also potentially caused by this current proposal which I've detailed below. I'll start by posting them now and continue to mull over potential solutions.

codablock committed:
Add applyIDMDChangesMutex to protect from parallel execution of InjectLabels and RemoveLabelsExcluded

This implements a temporary solution to protect agains parallel execution
of InjectLabels and RemoveLabelsExcluded, which might cause a different
form of the same deadlock as fixed by the previous commit.

A better solution should be implemented in the future, as mentioned in
the PR comment https://github.com/cilium/cilium/pull/18343#discussion_r778447874

Signed-off-by: Alexander Block <ablock84@gmail.com>

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -41,6 +41,11 @@ var (
 
+	// applyIDMDChangesMutex protects InjectLabels and RemoveLabelsExcluded from being run in parallel
+	//
+	// This mutex is a temporary solution until a better approach is implemented
```
nit: I don't think it's necessary to mention this. The solution is fine; it's not optimal but also unless we have immediate plans to implement a better approach, I don't think it really helps readers to reason about why this mutex exists. Temporary solutions will always live longer than you ever expect.
```suggestion
```

codablock replied:
I removed this comment and force-pushed

joestringer reviewed (commented):
The latest changes look good for mitigating the case (2) I raised above.

Case (1), `RemoveLabelsExcluded()` -> `removeLabelsFromIPs()` -> `updater.UpdateIdentities(...)` is still susceptible to a similar issue. I'm OK to either run tests & merge this PR as-is and then address that additional issue in a separate PR, or alternatively if you're up for trying to resolve that case as well in this PR then I would welcome it. This probably means releasing both (IDMD, IPIdentityCache) locks before the `UpdateIdentities()` calls inside `removeLabelsFromIPs()`, then after the policy updates were triggered, grab the IPIdentityCache lock again.

joestringer reviewed (approved):
There was only a non-functional change since the last bump, and the tests all previously passed. This is good as-is. As mentioned above, there's still one more similar case to cover but there's no need to address that directly in this PR, we can address that additional bug in a subsequent follow-up.
